### PR TITLE
Add incoming message handler for conference component

### PR DESCRIPTION
**Pull request by niuyunsi**

### CODE REVIEW

File: src/components/Conference.tsx [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+export class Conference extends React.Component<IConferenceProps, void> {
+    private connection: ConferenceConnection;
+    private peerConnections: { [id: string]: RTCPeerConnection };
+
+    constructor() {
+        super();
+        this.connection = this.props.connect();
+        this.connection.subscribe(this.handleIncomingMessage);
```
Comment (asinner) [inline src/components/Conference.tsx]: this.handleIncomingMessage.bind(this)

File: src/components/Conference.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    // TODO(yunsi): Complete display view.
+    public render() {
+        return (
+            <div />
+        )
+    }
+
+    public componentWillMount() {
```
Comment (asinner) [inline src/components/Conference.tsx]: componentWillUnmount?

File: src/components/Conference.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        this.leaveRoom()
+    }
+
+    private sendMessage(message: IConfOutgoingMessage) {
+        this.connection.publish(message);
+    }
+
+    private joinRoom(room: string) {
+        const message = {
```
Comment (asinner) [inline src/components/Conference.tsx]: Make a service file with functions for creating IConfOutgoingMessages

File: src/components/Conference.tsx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            type: 'Join',
+            room: room,
+        } as IConfOutgoingMessage;
+        this.sendMessage(message);
+    }
+
+    private leaveRoom() {
+        const message = {
```
Comment (asinner) [inline src/components/Conference.tsx]: same

File: src/components/Conference.tsx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+        if (peerConnection) {
+            peerConnection.createOffer(
+                sessionDescription => this.setLocalAndSendMessage(sessionDescription, 'Offer', id)
+            )
+        }
+    }
+
+    private createPeerConnectionById(id: string) {
+        try {
```
Comment (asinner) [inline src/components/Conference.tsx]: why does this need a try/catch?

File: src/components/Conference.tsx [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    private leaveRoom() {
+        const message = {
+            type: 'Bye',
+        } as IConfOutgoingMessage;
+        this.sendMessage(message);
+    }
+
+    private getUserMedia() {
+        navigator.mediaDevices.getUserMedia(userMediaConfig).then(stream => {
+            // TODO(yunsi): Send MediaStream to Local stream component.
```
Comment (asinner) [inline src/components/Conference.tsx]: for now, keep a reference to the local stream `this.stream = stream`

File: src/components/Conference.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+            peerConnection.createOffer(
+                sessionDescription => this.setLocalAndSendMessage(sessionDescription, 'Offer', id)
+            )
+        }
+    }
+
+    private createPeerConnectionById(id: string) {
+        try {
+            // TODO(yunsi): Add RTCPeerConnection config.
+            const peerConnection = new RTCPeerConnection({});
```
Comment (asinner) [inline src/components/Conference.tsx]: after creating peer connection, you have to call `peerConnection.addStream(this.stream)` with local stream

File: src/components/Conference.tsx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+            return peerConnection;
+        } catch (err) {
+            console.log('Can not create PeerConnection by id: ' + id);
+        }
+    }
+
+    private handleIceCandidate(event: RTCPeerConnectionIceEvent, id: string) {
+        if (event.candidate) {
+            const message = {
```
Comment (asinner) [inline src/components/Conference.tsx]: same

File: src/components/Conference.tsx [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+                candidate: event.candidate,
+                to: id,
+            } as IConfOutgoingMessage;
+            this.sendMessage(message);
+        }
+    }
+
+    private setLocalAndSendMessage(sessionDescription: RTCSessionDescription, type: string, id: string) {
+        const peerConnection = this.getPeerConnectionById(id);
+        const outgoingMessage = {
```
Comment (asinner) [inline src/components/Conference.tsx]: same

File: src/components/Conference.tsx [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        } as IConfOutgoingMessage
+
+        if (peerConnection) {
+            peerConnection.setLocalDescription(sessionDescription);
+            this.sendMessage(outgoingMessage);
+        }
+    }
+
+    private handleRemoteStreamAdded(event: MediaStreamEvent, id: string) {
+        // TODO(yunsi): Send MediaStream to Remote stream component.
```
Comment (asinner) [inline src/components/Conference.tsx:159]: For now keep a map of peer id -> stream

File: src/components/Conference.tsx [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            delete this.peerConnections[id];
+        }
+    }
+
+    private handleCandidateMessage(message: IConfIncomingMessageCandidate) {
+        const id = message.from;
+        const peerConnection = this.getPeerConnectionById(id);
+
+        if (peerConnection) {
+            peerConnection.addIceCandidate(message.candidate);
```
Comment (asinner) [inline src/components/Conference.tsx]: You have to make sure remote description is set before calling addIceCandidate

File: src/components/Conference.tsx [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        }
+    }
+
+    // NOTE(yunsi): When received an AddPeer event, conference will create a new PeerConnection and add it to the connection list.
+    private handleAddPeerMessage(message: IConfMessageAddPeer) {
+        const id = message.Id;
+        const peerConnection = this.createPeerConnectionById(id);
+
+        if (peerConnection) {
+            peerConnection.createOffer(
```
Comment (asinner) [inline src/components/Conference.tsx]: This should be conditional

File: src/components/Conference.tsx (hunk 1 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 6 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 10 above)
Comment (niuyunsi) [reply src/components/Conference.tsx:159]: Done

File: src/components/Conference.tsx (hunk 7 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 12 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 8 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 9 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 11 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 2 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 3 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 4 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    private remoteStreams: { [id: string]: MediaStream };
+    private localId: string;
+
+    constructor() {
+        super();
+        this.connection = this.props.connect();
+        this.connection.subscribe(this.handleIncomingMessage);
+        this.handleIncomingMessage.bind(this);
+        this.joinRoom(this.props.room);
+        this.getUserMedia();
```
Comment (asinner) [inline src/components/Conference.tsx]: There is a race condition here. You should only call `this.connection.subscribe` after getUserMedia has set the local stream

File: src/components/Conference.tsx [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    // NOTE(yunsi): When received an AddPeer event, conference will create a new PeerConnection and add it to the connection list.
+    private handleAddPeerMessage(message: IConfMessageAddPeer) {
+        const id = message.Id;
+        const peerConnection = this.createPeerConnectionById(id);
+
+        // NOTE(yunsi): When two clients both recieved an AddPeer event with the other client's id, they will do a compare to see who should create and send the offer.
+        if (this.localId.localeCompare(id)) {
+            peerConnection.createOffer(
+                sessionDescription => this.setLocalAndSendMessage(sessionDescription, 'Offer', id)
+            )
```
Comment (asinner) [inline src/components/Conference.tsx]: Add error handling TODO

File: src/components/Conference.tsx [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+        this.localId = message.Id;
+    }
+
+    // NOTE(yunsi): When received an AddPeer event, conference will create a new PeerConnection and add it to the connection list.
+    private handleAddPeerMessage(message: IConfMessageAddPeer) {
+        const id = message.Id;
+        const peerConnection = this.createPeerConnectionById(id);
+
+        // NOTE(yunsi): When two clients both recieved an AddPeer event with the other client's id, they will do a compare to see who should create and send the offer.
+        if (this.localId.localeCompare(id)) {
```
Comment (asinner) [inline src/components/Conference.tsx]: should be `this.localId.localeCompare(id) === 1` or `this.localId.localeCompare(id) === -1`

File: src/components/Conference.tsx [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+        const peerConnection = this.getPeerConnectionById(id);
+        let message;
+        if (type === 'Offer') {
+            message = createOutgoingMessageOffer(sessionDescription, id);
+        } else if (type === 'Answer') {
+            message = createOutgoingMessageAnswer(sessionDescription, id);
+        }
+
+        if (peerConnection && message) {
+            peerConnection.setLocalDescription(sessionDescription);
```
Comment (asinner) [inline src/components/Conference.tsx:152]: Add error handling TODO

File: src/components/Conference.tsx [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    // NOTE(yunsi): When received an Offer event, conference will set it as RemoteDescription and create an answer to the offer.
+    private handleOfferMessage(message: IConfIncomingMessageOffer) {
+        const id = message.from;
+        const peerConnection = this.getPeerConnectionById(id);
+
+        if (peerConnection) {
+            peerConnection.setRemoteDescription(message.sessionDescription);
+            peerConnection.createAnswer(
+                sessionDescription => this.setLocalAndSendMessage(sessionDescription, 'Answer', id)
+            )
```
Comment (asinner) [inline src/components/Conference.tsx]: Add error handling TODO

File: src/components/Conference.tsx [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+        }
+    }
+
+    // NOTE(yunsi): When received an Offer event, conference will set it as RemoteDescription and create an answer to the offer.
+    private handleOfferMessage(message: IConfIncomingMessageOffer) {
+        const id = message.from;
+        const peerConnection = this.getPeerConnectionById(id);
+
+        if (peerConnection) {
+            peerConnection.setRemoteDescription(message.sessionDescription);
```
Comment (asinner) [inline src/components/Conference.tsx]: Add error handling TODO

File: src/components/Conference.tsx [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+            peerConnection.createAnswer(
```
Comment (asinner) [inline src/components/Conference.tsx]: Does this need to be nested in setRemoteDescription?

File: src/components/Conference.tsx [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+            )
+        }
+    }
+
+    private handleAnswerMessage(message: IConfIncomingMessageAnswer) {
+        const id = message.from;
+        const peerConnection = this.getPeerConnectionById(id);
+
+        if (peerConnection) {
+            peerConnection.setRemoteDescription(message.sessionDescription);
```
Comment (asinner) [inline src/components/Conference.tsx]: Add error handling TODO

File: src/components/Conference.tsx [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+
+        delete this.peerConnections[id];
+        delete this.remoteStreams[id];
+    }
+
+    private handleCandidateMessage(message: IConfIncomingMessageCandidate) {
+        const id = message.from;
+        const peerConnection = this.getPeerConnectionById(id);
+
+        if (peerConnection && peerConnection.remoteDescription) {
```
Comment (asinner) [inline src/components/Conference.tsx:182]: We talked in person about handling this race condition

File: src/components/Conference.tsx (hunk 13 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 15 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 14 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 16 above)
Comment (niuyunsi) [reply src/components/Conference.tsx:152]: Done

File: src/components/Conference.tsx (hunk 21 above)
Comment (niuyunsi) [reply src/components/Conference.tsx:182]: Done

File: src/components/Conference.tsx (hunk 18 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 19 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 17 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 20 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

**Review by asinner — APPROVED** [review]

**Pull request merged**